WMF Framework/Remote Notifications/RemoteNotificationsAPIController.swift

```diff
@@ -116,25 +117,64 @@ struct RemoteNotificationsAPIController {
 
     public func markAsRead(_ notifications: Set<RemoteNotification>, completion: @escaping (Error?) -> Void) {
         let maxNumberOfNotificationsPerRequest = 50
+        let notifications = Array(notifications)
 
-        guard notifications.count <= maxNumberOfNotificationsPerRequest else {
-            // TODO: Split requests? 50 is the limit.
-            assertionFailure()
-            return
-        }
-        
-        
-
-        request(Query.markAsRead(notifications: notifications), method: .post) { (result: MarkReadResult?, _, _, error) in
-            if let error = error {
-                completion(error)
+        if notifications.count > maxNumberOfNotificationsPerRequest {
+            let split = notifications.chunked(into: maxNumberOfNotificationsPerRequest)
+            let group = DispatchGroup()
+            var markAsReadError: Error? = nil
+            var failed = 0
+            for notifications in split {
+                group.enter()
+                request(Query.markAsRead(notifications: notifications), method: .post) { (result: MarkReadResult?, _, _, error) in
+                    if let error = error {
+                        failed += 1
+                        markAsReadError = error
+                    }
+                    guard let result = result else {
+                        assertionFailure("Expected result; make sure MarkReadResult maps the expected result correctly")
+                        failed += 1
+                        markAsReadError = MarkReadError.noResult
+                        group.leave()
+                        return
+                    }
+                    if !result.succeeded {
+                        failed += 1
+                        markAsReadError = MarkReadError.unknown
+                    }
+                    if let error = result.error {
+                        failed += 1
+                        markAsReadError = error
+                    }
+                    group.leave()
+                }
             }
-            guard let result = result, result.succeeded else {
-                assertionFailure()
-                completion(MarkReadError.unknown)
-                return
+            group.notify(queue: DispatchQueue.global(qos: .default)) {
+                if failed > 1 {
+                    DDLogError("\(failed) of \(split.count) mark as read requests failed")
+                }
+                completion(markAsReadError)
+            }
+        } else {
+            request(Query.markAsRead(notifications: notifications), method: .post) { (result: MarkReadResult?, _, _, error) in
```

Comment:
it'd simplify things to allow this to follow the same code path as above, it looks like `chunked` works regardless of whether or not the chunk size is larger than the array size

Reply:
good idea 👍